osu.Game/Online/Multiplayer/MultiplayerClient.cs

```diff
@@ -50,80 +52,68 @@ private void apiStateChanged(ValueChangedEvent<APIState> state)
-                         .Build();
-
-            // this is kind of SILLY
-            // https://github.com/dotnet/aspnetcore/issues/15198
-            connection.On<MultiplayerRoomState>(nameof(IMultiplayerClient.RoomStateChanged), ((IMultiplayerClient)this).RoomStateChanged);
-            connection.On<MultiplayerRoomUser>(nameof(IMultiplayerClient.UserJoined), ((IMultiplayerClient)this).UserJoined);
-            connection.On<MultiplayerRoomUser>(nameof(IMultiplayerClient.UserLeft), ((IMultiplayerClient)this).UserLeft);
-            connection.On<int>(nameof(IMultiplayerClient.HostChanged), ((IMultiplayerClient)this).HostChanged);
-            connection.On<MultiplayerRoomSettings>(nameof(IMultiplayerClient.SettingsChanged), ((IMultiplayerClient)this).SettingsChanged);
-            connection.On<int, MultiplayerUserState>(nameof(IMultiplayerClient.UserStateChanged), ((IMultiplayerClient)this).UserStateChanged);
-            connection.On(nameof(IMultiplayerClient.LoadRequested), ((IMultiplayerClient)this).LoadRequested);
-            connection.On(nameof(IMultiplayerClient.MatchStarted), ((IMultiplayerClient)this).MatchStarted);
-            connection.On(nameof(IMultiplayerClient.ResultsReady), ((IMultiplayerClient)this).ResultsReady);
-
-            connection.Closed += async ex =>
-            {
-                isConnected.Value = false;
+        private readonly SemaphoreSlim connectionLock = new SemaphoreSlim(1);
 
-                Logger.Log(ex != null
-                    ? $"Multiplayer client lost connection: {ex}"
-                    : "Multiplayer client disconnected", LoggingTarget.Network);
+        public Task Disconnect() => disconnect(true);
 
-                if (connection != null)
-                    await tryUntilConnected();
-            };
+        protected async Task Connect()
+        {
+            cancelExistingConnect();
 
-            await tryUntilConnected();
+            await connectionLock.WaitAsync();
 
-            async Task tryUntilConnected()
+            try
             {
-                Logger.Log("Multiplayer client connecting...", LoggingTarget.Network);
+                await disconnect(false);
+
+                // this token will be valid for the scope of this connection.
+                // if cancelled, we can be sure that a disconnect or reconnect is handled elsewhere.
+                var cancellationToken = connectCancelSource.Token;
 
                 while (api.State.Value == APIState.Online)
                 {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    Logger.Log("Multiplayer client connecting...", LoggingTarget.Network);
+
                     try
                     {
-                        Debug.Assert(connection != null);
+                        // importantly, rebuild the connection each attempt to get an updated access token.
+                        connection = createConnection(cancellationToken);
```

Comment:
Won't this spam new connection instances in the "retry connecting" case? Are they getting cleaned up anywhere? It doesn't look like they are.

Reply:
You mean disposing? Probably just via finalizer, didn't really think much of it.

Reply:
Right. Looking at the sources it looks like `StopAsync()` and `DisposeAsync()` only differ by that the latter disposes the connection's inner `IServiceProvider`. Doesn't seem like a huge deal, but the latter seems more correct to call, and I don't know what the signalr packages might be keeping in there anyways.

Reply:
I don't think this requires further changes with the conversion of `StopAsync` to `DisposeAsync`.

Reply:
Not sure that's correct. Consider the `catch (Exception e)` branch; if that gets hit, the order of operations (simplified) is:

```csharp
await disconnect(false); // outside of while loop!

// while (api.State.Value == APIState.Online)
try
{
	connection = createConnection(cancellationToken);
	await connection.StartAsync(cancellationToken); // throws
}
catch (Exception e)
{
	await Task.Delay(5000, cancellationToken);
}

// while loop spins again, disconnect doesn't run
```

`connection` won't get cleaned up on that path as far as I can see? If `await disconnect(false)` was moved into the while loop, *then* it'd have a chance of getting cleaned up.

Reply:
I was presuming that an exception during connection means the connection has been closed. If not it sounds like a pretty bad API provided by signalr..

Reply:
Closed - sure, but disposed? - not so much given that this is supposed to be a reusable object, which this code isn't reusing.